# Activité d'algorithmique autour du binaire

Afin de pouvoir personnaliser ce notebook sans détruire le fichier initial, vous allez tout d'abord aller dans le menu **`File`** puis **`Make a copy...`**. 

Renommez le classeur en cliquant sur son nom `Binaire et ...` et par exemple en remplaçant `copy1` par  *travail_1* ou autre à la fin du nom de fichier.

Refermez alors l'onglet précédent en utilisant le menu **`File`** puis **`Close and Halt`**. 

En cas d'erreur de manipulation, vous venez de vous assurer la possibilité de recommencer à zéro à tout moment.

## Prise de contact avec le langage et l'environnement

Vous allez utiliser le langage Python. Il a l'avantage d'être très simple à prendre en main et concis au niveau de sa syntaxe. Il est également très proche du langage algorithmique naturel que nous utilisons en classe.

L'environnement que vous avez sous les yeux se nomme  Jupyter notebook. Il fonctionne dans un navigateur internet et permet de mélanger des lignes de code Python et du texte. 

Les cellules de code sont facilement reconnaissables, elles sont précédées de `In [ ]:`.
Ci-dessous se trouve une cellule de type ***code*** qui contient des instructions Python.

In [ ]:
a=3
x="du texte"
def g(x):
    return (1.5*x)
f=g(a)

On peut alterner ainsi des paragraphes de texte donnant des explications et des portions de programme.

Pour valider le ***texte*** d'une cellule après l'avoir modifiée, ou exécuter le ***code*** qu'elle contient, on doit saisir la combinaison de 2 touches **`Maj + Entrée`**.
C'est pour le moment tout ce que vous avez besoin de savoir.

La cellule **ci-dessus** étant activée (encadrée avec un bord gauche gras et bleu) par exemple en cliquant dessus ou en utilisant les flèches, appuyez sur **`Maj + Entrée`** pour exécuter son code.

Il vous a semblé que rien ne se passait. En réalité python a travaillé en mémoire.

Exécutez de la même façon la ligne de code ci-dessous.

In [ ]:
%whos

Vous avez sous les yeux les 4 objets python de différentes natures présents en mémoire.
- `int` signifie que `a` est un *entier*.
- `str` signifie que `x` est une *chaine (string) de caractères*. 
- `function` signifie que `g` est une *fonction*. **Notez que** le `x` interne à la fonction **n'existe plus** en mémoire après l'appel de la fonction.
- `float` signifie que `f` est un *flottant*.

python utilise ces différents types car les façons de stocker en mémoire, de numériser un entier et un flottant sont différentes.


<H3>Seulement du texte et du code ?</H3>

On peut, dans une cellule markdown insérer directement du <B>code HTML</B> !!!
<P>donc une image :
   ![By Juan Lacruz [CC BY-SA 3.0  (https://creativecommons.org/licenses/by-sa/3.0)], from Wikimedia Commons](https://upload.wikimedia.org/wikipedia/commons/thumb/a/a1/Rupicapra.jpg/320px-Rupicapra.jpg "Une chèvre et son cabri") 
<P>ou un tableau
<TABLE><TR><TD>colonne 1</TD><TD>Colonne 2</TD></TR>
<TR><TD>Machin</TD><TD>Truc</TD></TR>
</TABLE>
Et même des formules mathématiques $\sum_{i=1}^{n}{i} =\dfrac{n(n+1)}{2}$ en utilisant le langage de description $\LaTeX$.

## Le binaire et python 

### de binaire vers décimal
Écrivons 10101 et 1101 en décimale avant de généraliser :

![](http://b.rynhe.de.tioch.free.fr/image/bin_a_dec2.png "2 conversions")


Voici un algorithme de conversion à compléter : 
![](http://b.rynhe.de.tioch.free.fr/image/algo_bin_dec.png "algo de conversion")

et sa traduction en python dans la cellule ci-dessous :

In [19]:
def decim(bi):       # bi est une chaine de caractère
    """Transforme une chaine de caractère formée de 0 et de 1 en nombre décimal"""
    d = 0       # la variable locale (interne à la fonction decim ) d est initialisée à 0
    for k in bi:     # on traite tous les caractères 
        val = int(k) # on convertit le kième caractère en entier
        d = 2*d +val  # On ajoute … au double de d
    return d         # On renvoie d

In [36]:
decim('101001')

41

Attention, la chaine de caractère saisie doit correspondre à un nombre binaire, sinon la valeur renvoyée n'aura aucun sens.

C'est un excellent exercice que de fabriquer un programme python qui valide au fur et à mesure la compatibilité du texte saisi.

### De décimal vers binaire 
Écrivons 13 en binaire avant de généraliser :

![](http://b.rynhe.de.tioch.free.fr/image/dec_a_bin2.png "1 conversion")

L’écriture en binaire est constitué des restes successifs dans la répétition des divisions euclidiennes par 2.

On déduit l’algorithme à compléter d’une fonction de conversion suivant :
![](http://b.rynhe.de.tioch.free.fr/image/algo_dec_a_bin.png "algo de conversion")

Complétez à la main l'exécution de cet algorithme appliqué au nombre 41 en complétant au fur et à mesure le tableau suivant :
<TABLE border cellspacing=2 rules=all><TR> <TD>deci</TD><TD>41</TD><TD>…</TD><TD>…</TD><TD>…</TD><TD>…</TD><TD>…</TD><TD>…</TD><TD>…</TD><TD>…</TD></TR>
<TR> <TD>reste</TD><TD>…</TD><TD>…</TD><TD>…</TD><TD>…</TD><TD>…</TD><TD>…</TD><TD>…</TD><TD>…</TD><TD>…</TD></TR>
<TR> <TD>bin</TD><TD>…</TD><TD>…</TD><TD>…</TD><TD>…</TD><TD>…</TD><TD>…</TD><TD>…</TD><TD>…</TD><TD>…</TD></TR>
</TABLE>

Si ce n'est pas clair, n’hésitez pas à recommencer pour un autre nombre.

Voici, dans la cellule ci-dessous, la traduction en python :

In [37]:
def binaire(d):
    """Transforme un nombre décimal en binaire"""
    bi = ""       # la chaine de caractères bi est initialisée vide
    while d > 0:
        reste = d%2            # d%2 est le reste de la division de d par 2
        bi = str(reste) + bi   # on convertit reste en chaine et on l'ajoute à gauche de bi
        d = d // 2                # en python, // désigne la division euclidienne
    return bi

In [40]:
binaire(5)

'101'

## Le codage hexadécimal
Pour finir aujourd’hui, nous allons évoquer un dernier type de codage, qui constitue une alternative pratique au codage binaire. Il s’agit du codage hexadécimal, autrement dit en base seize.

### Pourquoi ce choix à priori bizarre ? 
Tout d’abord, parce que le codage binaire, est une écriture ni très économique, ni très lisible.

Pas très économique : pour représenter un nombre entre 1 et 256, il faut utiliser systématiquement huit chiffres.

Pas très lisible : parce qu’au delà de 3 caractères, notre cerveau n’arrive pas à lire « automatiquement ».

Alors, une alternative toute naturelle, est de représenter l’octet non comme huit bits (ce que nous avons fait jusque là), mais comme deux paquets de 4 bits (les quatre de gauche dits de poids fort, et les quatre de droite).

### Voyons cela de plus près. 
Avec 4 bits, nous pouvons coder 2⁴ = 16 nombres différents. 

En base seize, 16 nombres différents se représentent avec un seul caractère ou chiffre (de même qu’en base 10, dix nombres se représentent avec un seul chiffre).

Quels symboles choisir pour les chiffres ? Pour les dix premiers, on n’est pas allé chercher bien loin : on a recyclé les dix chiffres de la base décimale. Les dix premiers nombres de la base seize s’écrivent donc tout bêtement 0, 1, 2, 3, 4, 5, 6, 7, 8, et 9. 

Là, il nous manque encore 6 chiffres, pour représenter les nombres que nous écrivons en décimal 10, 11, 12, 13, 14 et 15. Plutôt qu’inventer de nouveaux symboles (ce qu’on aurait très bien pu faire), on a recyclé les premières lettres de l’alphabet. Ainsi, par convention, A vaut 10, B vaut 11, etc. jusqu’à F qui vaut 15.

On s’aperçoit que cette base hexadécimale permet une représentation très simple des octets du binaire par 2 "chiffres".

### Un exemple
Prenons un octet au hasard : 1001 1110.

Complètez les pointillés ci-dessous :


#### un voyage par étape
On commence par convertir ce nombre en écriture décimale en complétant ci-dessous. 

1001 1110₂ 	= ( 1 × 2⁷ + 0 × 2⁶ + 0 × 2⁵ + 1 × 2⁴ + 1 × 2³ + 1 × 2² + 1 × 2¹ + 0 × 2⁰ )₁₀ 
		= 128 + 16 + 8 + 4 + 2 = 158
        
De là, il faut passer à la base hexadécimale.

Pour passer à la base hexadécimale, on utilise la division sur les nombres entiers apprise en primaire.

Dans la division euclidienne de 158 par 16, le quotient est  9 et le reste est 14 .

Ainsi 158 = 9× 16 + 14 = 9× 16¹ + 14 × 16⁰ = 9E ₁₆  .

Le nombre 1001 1110₂ ou 158 en décimal s’écrit donc 9E en hexadécimal.

#### un aller direct
Une autre méthode consiste à faire le voyage direct du binaire vers l’hexadécimal. 
Avec l’habitude, c’est nettement plus rapide !

Séparons 1001 1110 en 1001 (4 bits de poids fort) et 1110 (4 bits de poids faible).

1001₂ = (1 × 2³ + 0 × 2² + 0 × 2¹ + 1 × 2⁰ )₁₀ = 9₁₀ = 9₁₆  

1110₂ = (2³  + 2²  + 2¹ + 0*2⁰)₁₀ = 14₁₀ = E₁₆ 
Le nombre s’écrit donc  9E  en hexadécimal. 

Le codage hexadécimal est très souvent utilisé quand on a besoin de représenter les octets individuellement, car dans ce codage, tout octet correspond à seulement deux chiffres de 0 à F.

La encore, nous allons demander un coup de main à Python pour convertir plus rapidement.

Exécutez la cellule ci-dessous :

In [43]:
b = bin(150)
h = hex(38)
d = 0b10011110
e = 0x2f
u = ord('a')
c = chr(97)

- Que font les fonctions `bin()`, `hex()`, `ord()` et `chr()` ? 
- De quel type sont les renvois de ces 4 fonctions ? 
- De quel type sont `d`, `e`, `u` et `c` ? 
- À quoi correspondent les 2 premiers caractères de `d` et `e` ? 
- Selon vous, qui de `b` ou de `d` utilise le plus d'espace mémoire ?

Pour répondre à toutes ces questions, réutilisons l'instruction `%whos`.

In [44]:
%whos

Variable   Type        Data/Info
--------------------------------
b          str         0b10010110
binaire    function    <function binaire at 0x00000225918F7598>
c          str         a
d          int         158
decim      function    <function decim at 0x00000225918F79D8>
e          int         47
h          str         0x26
u          int         97


Complétez :
- `b` et `h` sont des chaines de caractères
- la fonction `bin()` prend un nombre et le renvoie en binaire
- la fonction `hex()` prend un un nombre et le renvoie en hexadécimal
- `u` est un entier
- la fonction `ord()` prend un nombre et renvoie son codage en base 10
- `c` est un ...
- la fonction `char()` prend un code ... et renvoie ...
- `d` et `e` sont des entier 
- les 2 caractères `0b` permettent à python de ...
- les 2 caractères `0x` permettent à python de ...
- Selon moi `b` utilise ... de mémoire que `d` .

Pour chaque points de suspension que vous compléterez ci-dessous, créez, complétez une cellule de code python qui justifie la réponse :

- 1010 1010₂ =  20 en décimal = AA en hexadécimal


- A82₁₆ =  en décimal = … en binaire 


- 247 = … en binaire = … en hexadécimal


- En unicode, 0101 0101 code …

- En unicode, 26₁₆ code …


- En unicode, 65₁₀ code …

- En unicode, 65₁₆ code …

- En unicode, "a" est codé …

- Dans le langage Python … sert à dire qu’on commente le programme

# Ce que vous pouvez retenir en plus sur le langage Python

## Les opérations de base


Opération | algorithme | python
 ---:    |   :---  |  :---  
addition | 2 + 3 | 2 + 3
Soustraction | 12 - 5 | 12 - 5
Multiplication | 3 × 6 | 3 * 6
Division  | 7 / 2 | 7 / 2
Quotient de la division euclidienne | 7 div 2 ou div(7,2) | 7//2  
Reste de la division euclidienne | 7 mod 2 ou reste de 7 divisé par 2 | 7%2
puissance | $7^2$ | 7\*\*2
racine carrée | $\sqrt{2}$ ou sqrt(2) | math.sqrt(2) 

pour `math.sqrt`, il faut que `import math` ait été évalué une fois précédemment, peu importe quand.

## Les variables

- L'affectation des valeurs au variables se fait très naturellement au moyen du signe =
- L'affichage du contenu d'une variable se fait par l'instruction ***print ()***

In [ ]:
a=3
print(a)

## Les fonctions
Les fonctions permettent de "factoriser" des suites d'opérations et de les appeler en passant un ou plusieurs paramètres.

Ce sont des objets python excessivement pratiques.

In [ ]:
def f(x, y): # Ne pas oublier les :
    # il peut y avoir plusieurs lignes indentées formant un bloc d'instructions
    return (x**2-y**2) # On finit toujours une fonction par le renvoie d'un ou plusieurs objets

définie la fonction $f$ de paramètres $x$ et $y$ qui renvoie $x^2-y^2$.

On appelle son exécution ainsi :

In [ ]:
f(4, 5)

## La boucle pour

Les boucles *Répéter pour* se définissent au moyen de l'instruction ***for variable in range(nb_itérations) :***

Comme python est mathématicien, il commence avec le premier entier (0) et donc, *variable* prendra les valeurs de 0 à nb_itérations***-1***.

In [ ]:
for i in range(5): # Ne pas oublier les :
    print (i)      # Ne pas oublier l'indentation

## La boucle tant que

Les boucles *Tant que* se définissent au moyen de l'instruction ***while (condition) :***

In [ ]:
i=1
while(i<10):
    i=i+2
print(i)

# Pour la prochaine fois ...
ou pour les plus rapides

Convertir les hexadécimaux ci-dessous en décimal et **à la main** puis vérifier avec python:
- E7
- 7D
- 1010

Convertir les décimaux ci-dessous en binaire et **à la main** puis vérifier avec python:
- 357
- 1010
- 85

Convertir les binaires ci-dessous en hexadécimal et **à la main** puis vérifier avec python:
- 1011 0010
- 0011 1110
- 1010 1001


In [ ]:
# vérifiez ici

** Sauvegardez le document obtenu en utilisant le menu** `File` **puis** `Save and Checkpoint`.

Fermez ce document en utilisant le menu `File` puis `Close and Halt` qui fermera le présent onglet et videra la mémoire et les processus utilisés par python.